### PR TITLE
osd: Remove extra call to reg_next_scrub() during splits

### PR DESCRIPTION
**Pull request by dzafman**
Add assert() to catch this in the future

Fixes: http://tracker.ceph.com/issues/16474

Signed-off-by: David Zafman <dzafman@redhat.com>

### CODE REVIEW

**Comment by athanatos** [comment]
lgtm! adding needs-qa

**Comment by dzafman** [comment]
http://pulpito.ceph.com/dzafman-2016-09-22_17:11:48-rados-wip-zafman-testing-distro-basic-smithi/

There is a failure I'm looking at in the run above.  An example is job 432090.  I'm rerunning that job with my wip-zafman-fixes ceph-qa-suite branch.

File: src/osd/PG.cc [hunk 1]
```diff
@@ -3411,6 +3411,7 @@ void PG::reg_next_scrub()
   double scrub_min_interval = 0, scrub_max_interval = 0;
   pool.info.opts.get(pool_opts_t::SCRUB_MIN_INTERVAL, &scrub_min_interval);
   pool.info.opts.get(pool_opts_t::SCRUB_MAX_INTERVAL, &scrub_max_interval);
+  assert(scrubber.scrub_reg_stamp == utime_t());
```
Comment (robbat2) [inline src/osd/PG.cc]: Further up, utime_t() was changing to ceph_clock_now(cct). Should this be changed here as well?

File: src/osd/PG.cc [hunk 2]
```diff
@@ -3420,8 +3421,10 @@ void PG::reg_next_scrub()
 
 void PG::unreg_next_scrub()
 {
-  if (is_primary())
+  if (is_primary()) {
     osd->unreg_pg_scrub(info.pgid, scrubber.scrub_reg_stamp);
+    scrubber.scrub_reg_stamp = utime_t();
```
Comment (robbat2) [inline src/osd/PG.cc]: Again, should this be ceph_clock_now(cct) instead of utime_t()?

File: src/osd/PG.cc (hunk 1 above)
Comment (dzafman) [inline src/osd/PG.cc]: @rabbat2 No, this was to indicate that no time is set.

File: src/osd/PG.cc (hunk 2 above)
Comment (dzafman) [inline src/osd/PG.cc]: @robbat2 Same here this is supposed to indicate that no time is set.

Comment (dzafman) [inline src/osd/PG.cc]: @robbat2 I still see the following at top of master for the utime_t ctor:

utime_t() { tv.tv_sec = 0; tv.tv_nsec = 0; }

**Comment by robbat2** [comment]
We should add in some docs that make it more clear why the unreg scrub uses zero as the value for scrub_reg_stamp. 5e44040e8528bff06cc0a5a3f3293ab146e0e4e1 is what prompted me to ask if ceph_clock_now was more correct.

**Comment by dzafman** [comment]
The _scrub_reg_stamp_  contents is a stashed value when a scrub is registered in reg_next_scrub() that is only used to allow unreg_next_scrub() to find and remove that registration later.  It is not the same as last_scrub_stamp or anything that can be seen outside of the OSD.  I set it to utime_t() to indicate that that scrub isn't registered, so we can assert if you try to remove a scrub twice.  That value is the same as what is initialized by the constructor of PG::Scrubber.

**Comment by dzafman** [comment]
Only one failure:  
79/146 Test  #86: unittest_erasure_code_plugin_jerasure ...***Exception: SegFault  0.13 sec

**Pull request merged**